lib/semantics/tools.cc

```diff
@@ -226,4 +227,25 @@ const Symbol *FindExternallyVisibleObject(
     return nullptr;
   }
 }
+
+bool ExprHasTypeCategory(const evaluate::GenericExprWrapper &expr,
+    const common::TypeCategory &type) {
+  auto dynamicType{expr.v.GetType()};
+  return dynamicType.has_value() && dynamicType->category == type;
+}
+
+void CheckScalarLogicalExpr(
+    const parser::Expr &expr, parser::Messages &messages) {
+  // TODO: should be asserting that typedExpr is not null
+  if (expr.typedExpr == nullptr) {
+    return;
+  }
+  // TODO: Whence IsArray()?
```

Comment:
You could add IsArray() to ExpressionBase or we could continue using `Rank() > 0`. I don't think we need the TODO.

Reply:
I'll stick with Rank for now until I see how assumed rank is represented.